Lyrics Improvements

kgarner7 opened the pull request:
- Make the settings text actually consistent with behavior
- Add metadata (artist/track name) for fetched tracks
- Add ability to remove incorrectly fetched lyric
- Add lyric fetch cache; save the last 10 fetches
- Add ability to change offset in full screen, add more comments

Sorry for another one of these so close.
Todo (?):
- [x] Add spinner when fetching? Not sure what counts as fetching; Jellyfin fetch? RPC Fetch? If only the first, this is easy, otherwise you'll need either a timeout or an ACK for no response...
- [x] Add the ability to manually search for lyrics (this would require a decent amount of work)...

kgarner7 committed:
Lyrics Improvements

- Make the settings text actually consistent with behavior
- Add metadata (artist/track name) for fetched tracks
- Add ability to remove incorrectly fetched lyric
- Add lyric fetch cache; save the last 10 fetches
- Add ability to change offset in full screen, add more comments

jeffvli commented:
After reviewing the implementation again, I believe it can be greatly simplified especially if we want to support loading indicators as well as manual search.

The current implementation is a three way back and forth process that follows some of the existing patterns in the app:
  1. When song changes, send a fetch request to the main process from the renderer
  2. Main process begins the fetch
  3. Listener function on the renderer receives the output of the fetch from the main process and then we can perform actions on it

This works but we don't have a way to easily check when the function is actually running unless we were to add an additional listener for that.

After a bit of research, it turns out you can return a promise from the main process using [ipcRenderer.invoke and ipcMain.handle](https://stackoverflow.com/questions/45148110/how-to-add-a-callback-to-ipc-renderer-send). Using this method we can now fetch the lyrics on demand.

If you don't mind, I can rewrite the lyric fetching process and then we can work on the other features after.

jeffvli committed:
Update lyric fetcher return types

jeffvli committed:
Replace main lyrics listener with promise handler

jeffvli committed:
Set lyrics scrollarea to max height by default

jeffvli committed:
Move all lyrics fetching logic into query

jeffvli committed:
Move scroll container to wrap lyrics only

jeffvli committed:
Add fade in animation for lyrics container

jeffvli commented:
I pushed some changes to a [separate branch](https://github.com/jeffvli/feishin/commits/lyrics-enhancements).

- Replaces the original lyric fetch and listener pattern
- Moves all lyric fetching logic into the query hook to support loaders and use built-in caching via react-query
  - We can now potentially remove the in-memory caching in the main process 

Can you review and let me know your thoughts? If you're ok with it, I'll push it to your branch or you can just merge it into yours.

TODO:
- [x] Add manual search
  - [x] `MAIN` For each source, if possible by the source, add a function that returns an array of search results. Parameters for the function should include "name" and "artist"
  - [x] `MAIN` Add new `invoke` and `handle` functions to fetch the search results
  - [x] `RENDERER` Add new query that fetches the lyric search results
  - [x] `RENDERER` Add modal form UI with inputs for "name" and "artist" pre-filled with existing song data
- [ ] ~~Persist overridden song lyrics (tbd, need to determine best way to do this)~~ This will come in a later PR

kgarner7 committed:
Include lyric simplification, restore clear button

- merges lyric simplification
- restores metadata/clear

kgarner7 commented:
Looks good! I merged it in, and brought back the metadata/clear options.

jeffvli committed:
Fix animation transition on fullscreen player open

jeffvli committed:
Merge remote-tracking branch 'origin/development' into lyrics-enhancements

jeffvli committed:
Catch error on jellyfin query fail

jeffvli committed:
Add alternative lyrics format parser

- Many synced lyrics on NetEase are broken due to not being the standard lrc format

jeffvli committed:
Add lyric search functions and query

jeffvli committed:
Add initial lyrics search UI

jeffvli committed:
Fix ref issue

jeffvli committed:
Update fetchers to support search

jeffvli committed:
Add lyric search selection and override

jeffvli committed:
Miscellanous fixes

jeffvli committed:
Improve accessibility of actions bar

jeffvli committed:
Fix queue container layout

jeffvli committed:
Fix override on song change

jeffvli committed:
Add remaining lyric actions functionality

jeffvli committed:
Require only single query for search

jeffvli committed:
Disable action buttons when no song selected

jeffvli committed:
Fix provider text

jeffvli committed:
Allow override on songs without default found lyrics

jeffvli committed:
Improve lyrics match with scored searches

jeffvli committed:
Adjust header, input, and modal styles

jeffvli committed:
Fix id return type on netease fetcher

jeffvli committed:
Fix broken layout on lyrics scroll

- App layout would break when transitioning into the full-screen due to scrollIntoView
- Replace scroll action with scrollTop implementation

jeffvli committed:
Hide search / clear actions on web

jeffvli committed:
Adjust vertical padding on lyrics display

jeffvli committed:
Fix overlay scrollbar

jeffvli committed:
Add tooltip to offset input

jeffvli committed:
Add updated scroll container to unsync lyrics
